### PR TITLE
[7.x] [WIP] Switch from SwiftMailer to Symfony Mailer

### PR DESCRIPTION
**Pull request by jbrooksuk**
Looking to replace SwiftMailer with Symfony's new [Mailer](https://symfony.com/doc/current/mailer.html) component. Useful slides at https://speakerdeck.com/fabpot/mailer

**To-do:**

- [x] Replace core SwiftMailer implementation
- [ ] Replace `SesTransport` with https://github.com/symfony/amazon-mailer and `MailgunTransport` with https://github.com/symfony/mailgun-mailer
- [x] Update tests

**Thoughts / issues:**

- `sender` does not allow for named senders. I'm wondering if we can pass through a `NamedAddress` though - but this means that the second parameter is unused.
- Can we implement a DSN system for Mail too?

**What does this break?**

- Any mention of `Swift`, i.e. I've renamed the `getSwiftMessage` method to `getSymfonyEmail` to reflect what we're getting back.
- SwiftMailer plugins will no longer work.
- `send()` currently returns the amount of addresses the email was sent to, but that's now no longer the case
- `Message@getFrom` no longer returns an array of sender email / name, instead it's formatted as `Name <Email>`
- When sending an email fails, you get an array of addresses which the email couldn't be sent to, this has currently been removed.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Is the mailer component stable in Symfony 4.3?

**Comment by jbrooksuk** [comment]
Fabien removed the `experimental` tag from the classes, so I guess so? https://github.com/symfony/mailer/commit/53d22876f92182be4487d87325ec40fbc1931268

**Comment by GrahamCampbell** [comment]
Yeh, looks like it's only marked as stable in 4.4 and 5.0, so if Laravel uses ^4.3, it may find stuff breaks in 4.4.0. I d'no what symfony's intention is, regarding further refactoring in 4.4.

**Comment by jbrooksuk** [comment]
I've just set this branch up in a test project and it's sending emails!

**Comment by browner12** [comment]
Can you summarize the benefits of Symfony over Swift?  From my brief look at the slides the best I could glean is a smaller object, and dynamic vs fixed headers.

**Comment by jbrooksuk** [comment]
> Can you summarize the benefits of Symfony over Swift? From my brief look at the slides the best I could glean is a smaller object, and dynamic vs fixed headers.

Fabien has some slides at https://speakerdeck.com/fabpot/mailer which kind of explain more of the benefits.

**Comment by GrahamCampbell** [comment]
I major benafit is the native HTTP adapters symfony mailer provides. No more hacking around or needing lots of vendor SDKs.

**Comment by dinhquochan** [comment]
So, some packages depend on `guzzlehttp/guzzle` will be switch to `symfony/http-client` cuz `MailgunTransport` or `SesTransport`?

**Comment by jbrooksuk** [comment]
If the Symfony packages require that, then yes.

**Comment by taylorotwell** [comment]
How about attachments and inline images? Do they work the same?

**Comment by GrahamCampbell** [comment]
I think the concept of Laravel 5 mail transports isn't needed anymore surely? Just symfony mailer?

**Comment by jbrooksuk** [comment]
> I think the concept of Laravel 5 mail transports isn't needed anymore surely? Just symfony mailer?

Apart from the `Array` and `Log` transports, I think.

**Comment by jbrooksuk** [comment]
> How about attachments and inline images? Do they work the same?

I need to test sending them, but the abstraction was replaced, so the Laravel API should be the same.

**Comment by jbrooksuk** [comment]
Okay, I've checked out attachments this morning and they're sending just fine:

image

However, when looking at the `embed` methods these still need some work. The images I'm embedding don't seem to be rendering inline in the email clients.

image

I'm returning the CID from the `embed` method (which is a bit different now). Will push the changes.

**Comment by jbrooksuk** [comment]
FWIW, the embedded image is being attached, it's just not showing inline.

**Comment by barryvdh** [comment]
Besides the code benefits, it would be beneficial to include this in 6.0 because of SemVer starting there. And Symfony will probably stop supporting Swiftmailer and focus on the new Mailer component.

Do you need help with the inline images?

**Comment by jbrooksuk** [comment]
> Do you need help with the inline images?

@barryvdh if you have the time, that'd be great. If not, I'll pick this up again this week.

**Comment by driesvints** [comment]
FYI: I still think it's best we hold this off until Laravel 7 because it's still experimental. It'll be stable in Symfony 4.4 and 5.0.

**Comment by barryvdh** [comment]
Hmm yeah the Mail/Message class is pretty tied to the Swiftmailer/Symfony object, so it would be difficult to seperate it or do it later without breaking BC. But experimental could be tricky.
I suppose Laravel 6.0 is planned before November (when 4.4) releases? Because an LTS would be a nice starting point for the new 'major' version ;)

**Comment by driesvints** [comment]
@barryvdh Laravel 6 will be released end of this month or early September.

**Review by darielopper — APPROVED** [review]

**Comment by GrahamCampbell** [comment]
@dramosv90 this PR is definitely not ready yet. ;)

**Review by rbuzzibarrios — APPROVED** [review]

**Comment by taylorotwell** [comment]
Just closing this for now since there have been big changes in the mailer.

**Comment by barryvdh** [comment]
https://symfony.com/blog/the-end-of-swiftmailer

> Now comes the time to officially end the maintenance of Swiftmailer in favor of Symfony Mailer. I've set the end of maintenance date to the end of November 2021, at the same time as the Symfony 5.4 LTS and 6.0 release.

Should this be re-opened/re-evaluated?

**Comment by driesvints** [comment]
@barryvdh yeah I guess we should have another look. Thanks for mentioning that.
